lib/ace/ext/papeeria/papeeria_latex_highlight_rules_test.js

```diff
@@ -0,0 +1,176 @@
+
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_NUMBERED_STATE].end = "\\end{equation}";
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_STATE].end = "\\end{equation*}";
+    mathConstants[MATH_TEX_INLINE_STATE].end = "$";
+    mathConstants[MATH_TEX_DISPLAYED_STATE].end = "$$";
+    mathConstants[MATH_LATEX_INLINE_STATE].end = "\\)";
+    mathConstants[MATH_LATEX_DISPLAYED_STATE].end = "\\]";
+
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_NUMBERED_STATE].length = 4;
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_STATE].length = 4;
+    mathConstants[MATH_TEX_INLINE_STATE].length = 1;
+    mathConstants[MATH_TEX_DISPLAYED_STATE].length = 1;
+    mathConstants[MATH_LATEX_INLINE_STATE].length = 1;
+    mathConstants[MATH_LATEX_DISPLAYED_STATE].length = 1;
+
+    module.exports = {
+        "test: basic testing": function() {
+            var rules = new PapeeriaLatexHighlightRules().getRules();
+            var tokenizer = new Tokenizer(rules);
+            var basicString = " x^2 \\text{hey} \\alpha \\$ % this is a comment";
+            for (var state in mathConstants) {
+                var stateConstants = mathConstants[state];
+                var testString = stateConstants.start + basicString;
+                var result = tokenizer.getLineTokens(testString);
+                var tokens = result.tokens;
+                var state = result.state;
+                assert.ok(state.length);
+                for (var i = stateConstants.length; i < tokens.length; i++) {
+                    assert(isType(tokens[i], EQUATION_TOKEN_TYPE), state);
+                }
+            }
+        },
+
+        "test: mismatching ends": function() {
+            var rules = new PapeeriaLatexHighlightRules().getRules();
+            var tokenizer = new Tokenizer(rules);
+
+            for (var startState in mathConstants) {
+                var startConstants = mathConstants[startState];
+                for (var endState in mathConstants) {
+                    var endConstants = mathConstants[endState];
+                    if (endConstants.end.startsWith(startConstants.end)) {
+                        continue;
+                    }
+
+                    var testString = startConstants.start + " " + endConstants.end;
+                    var result = tokenizer.getLineTokens(testString);
+                    assert.equal(result.state[result.state.length - 1], startState);
+                }
+            }
+        },
+
+        "test: nested equation starts": function() {
+            var rules = new PapeeriaLatexHighlightRules().getRules();
+            var tokenizer = new Tokenizer(rules);
+
+            for (var initState in mathConstants) {
+                var initStateConstants = mathConstants[initState];
+                for (var innerState in mathConstants) {
+                    var innerStateConstants = mathConstants[innerState];
+                    if (innerStateConstants.start.startsWith(initStateConstants.end)) {
```

Comment:
This code is meant to test if equation start inside equation state changes state or not (it shouldn't). It is done by trying to parse one of the equation state starts, while already inside another equation state (second argument in `tokenizer.getLineTokens`). For instance, it will test, whether `\begin{equation}` (`math.environment.displayed.numbered` state start) parsed inside `math.latex.displayed` state (the one delimited with `\[` and `\]`) changes the state.

However, consider another case, where we try to parse `$$` inside `math.tex.inline` state (the one delimited with `$`). In this case, `initStateConstants.end` is actually a prefix of `innerStateConstants.start`, and so `innerStateConstants.start` will actually be parsed as an end of `math.tex.inline` state, and this is the correct behaviour.